stl/inc/algorithm

```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
     // not parallelized as benchmarks show it isn't worth it
     return _STD shift_right(_First, _Last, _Pos_to_shift);
 }
+
+#if _HAS_CXX23 && defined(__cpp_lib_concepts)
+namespace ranges {
+    template <class _It, class _Se>
+    constexpr _It _Shift_left_unchecked(_It _First, const _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
+        if (_Pos_to_shift <= 0) {
+            _RANGES advance(_First, _Last);
+            return _First;
+        }
+
+        auto _Start_at = _First;
+
+        if constexpr (sized_sentinel_for<_Se, _It>) {
+            if (_Pos_to_shift >= _Last - _First) {
+                return _First;
+            }
+            _RANGES advance(_Start_at, _Pos_to_shift);
+        } else {
+            for (; 0 < _Pos_to_shift; --_Pos_to_shift) {
+                if (_Start_at == _Last) {
+                    return _First;
+                }
+                ++_Start_at;
+            }
+        }
+
+        return _RANGES _Move_unchecked(_Start_at, _Last, _First).out;
+    }
+
+    template <permutable _It, sentinel_for<_It> _Se>
+    constexpr subrange<_It> shift_left(_It _First, _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
+        _Adl_verify_range(_First, _Last);
+        auto _Result = _First;
+        _Seek_wrapped(_Result, _Shift_left_unchecked(_Get_unwrapped(_First), _Get_unwrapped(_Last), _Pos_to_shift));
+        return {_First, _Result};
+    }
+
+    template <forward_range _Rng>
+        requires permutable<iterator_t<_Rng>>
+    constexpr borrowed_subrange_t<_Rng> shift_left(_Rng&& _Range, range_difference_t<_Rng> _Pos_to_shift) {
+        auto _First  = _RANGES begin(_Range);
+        auto _Result = _First;
+        _Seek_wrapped(_Result, _Shift_left_unchecked(_STD move(_Get_unwrapped(_Result)), _Uend(_Range), _Pos_to_shift));
+        return {_First, _Result};
+    }
+
+    template <class _It, typename _Se>

```

Comment:
Style: We follow an ancient convention of `class` instead of `typename` for template parameters in product code (although I personally prefer `typename`).